abd-clam/src/cakes/knn/sieve_v1.rs

```diff
@@ -29,257 +27,154 @@ where
+
+        let (insiders, non_insiders) = grains.split_at_mut(i);
+        if non_insiders.is_empty() {
+            grains_to_hits(data, query, &mut hits, insiders);
+            break;
         }
-    }
-
-    /// One-time computation of `grains.`
-    pub fn initialize_grains(&mut self) {
-        let root = self.tree.root();
-        let distance = root.distance_to_instance(self.tree.data(), self.query);
-        let grain = Grain::new(root, distance + root.radius, root.cardinality);
-        self.grains = vec![grain];
-    }
-
-    /// Returns whether search is complete, i.e., whether hits
-    /// contains the k-nearest neighbors.
-    const fn is_refined(&self) -> bool {
-        self.is_refined
-    }
 
-    /// One iteration of the refinement step.
-    pub fn refine_step(&mut self) {
-        // Determines the index of the grain such that if we only retain that grain
-        // and every closer grain, we are guaranteed to have at least k points.
-        // Threshold is the distance of the furthest point in the grain.
-        let i = Grain::partition_kth(&mut self.grains, self.k);
-        let ith_grain = &self.grains[i];
-        let threshold = ith_grain.d;
-
-        // Filters hits by being outside the threshold.
-        while !self.hits.is_empty()
-            && self
-                .hits
-                .peek_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"))
-                .1
-                .number
-                > threshold
-        {
-            self.hits
-                .pop_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"));
+        let (small_insiders, insiders) = insiders
+            .iter_mut()
+            .map(|&mut g| g)
+            .partition::<Vec<_>, _>(|g| g.is_small(k));
+        grains_to_hits(data, query, &mut hits, &small_insiders);
+
+        let (leaves, non_leaves) = non_insiders
+            .iter_mut()
+            .filter(|g| !g.is_outside(threshold))
+            .map(|&mut g| g)
+            .partition::<Vec<_>, _>(|g| g.is_leaf);
+
+        // If there are no non-leaves then the search is complete.
+        if non_leaves.is_empty() {
+            grains_to_hits(data, query, &mut hits, &insiders);
+            grains_to_hits(data, query, &mut hits, &leaves);
+            break;
         }
 
-        // Partition into insiders and straddlers
-        #[allow(clippy::iter_with_drain)] // clippy is wrong in this case
-        let (mut insiders, straddlers): (Vec<_>, Vec<_>) = self
-            .grains
-            .drain(..)
-            .filter(|g| !Grain::is_outside(g, threshold))
-            .partition(|g| Grain::is_inside(g, threshold));
-
-        // Distinguish between those insiders we won't partition further and those we will
-        // Add instances from insiders we won't further partition to hits
-        let (small_insiders, big_insiders): (Vec<_>, Vec<_>) = insiders
+        grains = non_leaves
```

Comment:
In the same vein as my previous comment, what about just adding all the leaves to hits? I.e., no hit makes it to the next iteration. Very non-rigorously, if you're a cluster and you've survived long enough for us to partition you down to a leaf, you must not suck too much lest you'd have been eliminated by now.